opentelemetry-api/src/opentelemetry/trace/span.py

```diff
@@ -157,28 +159,59 @@ class SpanContext:
         is_remote: True if propagated from a remote parent.
     """
 
-    def __init__(
-        self,
+    def __new__(
+        cls,
         trace_id: int,
         span_id: int,
         is_remote: bool,
         trace_flags: "TraceFlags" = DEFAULT_TRACE_OPTIONS,
         trace_state: "TraceState" = DEFAULT_TRACE_STATE,
-    ) -> None:
+    ) -> "SpanContext":
         if trace_flags is None:
             trace_flags = DEFAULT_TRACE_OPTIONS
         if trace_state is None:
             trace_state = DEFAULT_TRACE_STATE
-        self.trace_id = trace_id
-        self.span_id = span_id
-        self.trace_flags = trace_flags
-        self.trace_state = trace_state
-        self.is_remote = is_remote
-        self.is_valid = (
-            self.trace_id != INVALID_TRACE_ID
-            and self.span_id != INVALID_SPAN_ID
+
+        is_valid = trace_id != INVALID_TRACE_ID and span_id != INVALID_SPAN_ID
+
+        return tuple.__new__(
+            cls,
+            (trace_id, span_id, is_remote, trace_flags, trace_state, is_valid),
         )
 
+    @property
+    def trace_id(self) -> int:
+        return self[0]
+
+    @property
+    def span_id(self) -> int:
+        return self[1]
+
+    @property
+    def is_remote(self) -> bool:
+        return self[2]
+
+    @property
+    def trace_flags(self) -> "TraceFlags":
+        return self[3]
+
+    @property
+    def trace_state(self) -> "TraceState":
+        return self[4]
+
+    @property
+    def is_valid(self) -> bool:
+        return self[5]
+
+    def __getattr__(self, *args: str) -> None:
+        raise TypeError
+
+    def __setattr__(self, *args: str) -> None:
+        raise TypeError
```

Comment:
How strict are we in following that we should never crash the user's application? Should be maybe do nothing and log a warning instead?

Reply:
I agree. Logging a warning is definitely a better way to go.